aten/src/ATen/native/cpu/IndexKernel.cpp

```diff
@@ -0,0 +1,125 @@
+#include <ATen/native/Indexing.h>
+
+#include <cmath>
+#include <iostream>
+#include <ATen/Dispatch.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/Parallel.h>
+#include <ATen/cpu/vec256/vec256.h>
+
+namespace at { namespace native {
+namespace {
+
+using namespace vec256;
+
+struct Indexer {
+  Indexer(int64_t num_indexers, char** indexers, const int64_t* indexer_strides,
+          IntList original_sizes, IntList original_strides)
+    : num_indexers(num_indexers)
+    , indexers(indexers)
+    , indexer_strides(indexer_strides)
+    , original_strides(original_strides.data())
+    , original_sizes(original_sizes.data()) {
+    AT_ASSERT(original_strides.size() == num_indexers);
+    AT_ASSERT(original_sizes.size() == num_indexers);
+  }
+
+  int64_t num_indexers;
+  char** indexers;
+  const int64_t* indexer_strides;
+  const int64_t* original_strides;
+  const int64_t* original_sizes;
+
+  int64_t get(int64_t idx) {
+    int64_t offset = 0;
+    for (int j = 0; j < num_indexers; j++) {
+      int64_t value = *(int64_t*)&indexers[j][idx * indexer_strides[j]];
+      int64_t size = original_sizes[j];
+      if (value < -size || value >= size) {
+        AT_ERROR("index ", value, " is out of bounds for dim with size ", size);
+      }
+      if (value < 0) {
+        value += size;
+      }
+      offset += value * original_strides[j];
+    }
+    return offset;
+  }
+};
+
+static bool is_constant_index(int ntensor, const int64_t* strides) {
+  AT_ASSERT(ntensor >= 3);
+  for (int arg = 2; arg < ntensor; arg++) {
+    if (strides[arg] != 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename scalar_t, typename func_t>
+void cpu_index_kernel(TensorIterator& iter, IntList index_size, IntList index_stride,
+                      const func_t& f, bool serial_execution=false)
+{
+  auto loop = [&](int ntensor, char** data, const int64_t* strides, int64_t n) {
+    auto indexer = Indexer(ntensor - 2, &data[2], &strides[2], index_size, index_stride);
+    char* dst = data[0];
+    char* src = data[1];
+    if (is_constant_index(ntensor, strides)) {
+      // specialization for when every element uses the same index
+      int64_t offset = indexer.get(0);
+      if (strides[0] == sizeof(scalar_t) && strides[1] == sizeof(scalar_t)) {
+        for (int64_t i = 0; i < n; i++) {
```

Comment:
It looks like this loop is exactly the same as the one in the else statement. Is that because some compiler magic will happen here as it recognise the condition? Or some of the stride lookups should be removed in this loop?

Reply:
I think this is a trick in numpy as well, which helps the compiler produce more effluent code, but I'm not 100% sure

Reply:
It's a trick to encourage auto-vectorization. The compiler propagates the assumptions from the if condition to the body. So even though the bodies contain the same C++ code they get compiled to different instructions because the propagated assumptions are different.